### PR TITLE
Bugfix: Lock access to creating and removing directories since it's not thread-safe

### PR DESCRIPTION
**Pull request by eliran-stratoscale**
It seems that different threads use the _drafts object, so the _directoryExists boolean must be guarded. Otherwise, there are situations in which a thread thinks the directory exists (while it does not really exist), tries to write to it and gets a boost exception (this happens once every few times when running the UT).

### CODE REVIEW

File: cpp/Osmosis/ObjectStore/DirectoryValidator.h [hunk 1]
```diff
@@ -0,0 +1,49 @@
+#ifndef __OSMOSIS_OBJECT_STORE_DIRECTORY_VALIDATOR_H__
+#define __OSMOSIS_OBJECT_STORE_DIRECTORY_VALIDATOR_H__
+
+#include <boost/filesystem.hpp>
+#include <mutex>
+
+namespace Osmosis {
+namespace ObjectStore {
```
Comment (shlomi-stratoscale) [inline cpp/Osmosis/ObjectStore/DirectoryValidator.h]: inconsistent code convention

File: cpp/Osmosis/ObjectStore/DirectoryValidator.h [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+                if ( boost::filesystem::is_directory( _path ) )
+                        return;
+                const bool ret = boost::filesystem::create_directories( _path );
+                ASSERT( ret );
+		( void )ret;
+	}
+
+	void erase()
+	{
+                std::lock_guard< std::mutex > lock( _directoryExistsLock );
```
Comment (shlomi-stratoscale) [inline cpp/Osmosis/ObjectStore/DirectoryValidator.h]: indentation issue

File: cpp/Osmosis/ObjectStore/DirectoryValidator.h [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+                if ( _directoryExists )
+                        return;
+                _directoryExists = true;
+                if ( boost::filesystem::is_directory( _path ) )
+                        return;
+                const bool ret = boost::filesystem::create_directories( _path );
+                ASSERT( ret );
+		( void )ret;
```
Comment (shlomi-stratoscale) [inline cpp/Osmosis/ObjectStore/DirectoryValidator.h]: indentation issue

File: cpp/Osmosis/ObjectStore/DirectoryValidator.h [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+{
+public:
+	DirectoryValidator( const boost::filesystem::path & path) :
+		_path( path ),
+		_directoryExists( false )
+	{ }
+
+	void makeSureExists()
+	{
+                std::lock_guard< std::mutex > lock( _directoryExistsLock );
```
Comment (shlomi-stratoscale) [inline cpp/Osmosis/ObjectStore/DirectoryValidator.h]: well, just a thought: a mutex is not a cheap operation... i would add "if ( _directoryExists )" here as well (yes, you have to repeat it again after the lock). so maybe it's even better to split it into a second internal function

File: cpp/Osmosis/ObjectStore/DirectoryValidator.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                if ( _directoryExists )
+                        return;
+                _directoryExists = true;
+                if ( boost::filesystem::is_directory( _path ) )
+                        return;
+                const bool ret = boost::filesystem::create_directories( _path );
+                ASSERT( ret );
```
Comment (shlomi-stratoscale) [inline cpp/Osmosis/ObjectStore/DirectoryValidator.h]: so why is this an assert?

File: cpp/Osmosis/ObjectStore/DirectoryValidator.h [hunk 6]
```diff
@@ -0,0 +1,49 @@
+#ifndef __OSMOSIS_OBJECT_STORE_DIRECTORY_VALIDATOR_H__
+#define __OSMOSIS_OBJECT_STORE_DIRECTORY_VALIDATOR_H__
+
+#include <boost/filesystem.hpp>
+#include <mutex>
+
+namespace Osmosis {
+namespace ObjectStore {
+
+class DirectoryValidator
```
Comment (shlomi-stratoscale) [inline cpp/Osmosis/ObjectStore/DirectoryValidator.h]: name too generic. how about "MakeSureDirectoryExists", or "DirectoryExistance", or even "MakeDirectory". DirectoryValidator sounds like there is something about it's structure that requires validating, but doesn't state why, so the code using this class can not be understood without reading this file...

File: cpp/Osmosis/ObjectStore/DirectoryValidator.h (hunk 1 above)
Comment (eliran-stratoscale) [inline cpp/Osmosis/ObjectStore/DirectoryValidator.h]: Fixed

File: cpp/Osmosis/ObjectStore/DirectoryValidator.h (hunk 6 above)
Comment (eliran-stratoscale) [inline cpp/Osmosis/ObjectStore/DirectoryValidator.h]: Fixed

File: cpp/Osmosis/ObjectStore/DirectoryValidator.h (hunk 3 above)
Comment (eliran-stratoscale) [inline cpp/Osmosis/ObjectStore/DirectoryValidator.h]: Fixed

File: cpp/Osmosis/ObjectStore/DirectoryValidator.h (hunk 2 above)
Comment (eliran-stratoscale) [inline cpp/Osmosis/ObjectStore/DirectoryValidator.h]: Fixed

File: cpp/Osmosis/ObjectStore/DirectoryValidator.h (hunk 5 above)
Comment (eliran-stratoscale) [inline cpp/Osmosis/ObjectStore/DirectoryValidator.h]: What else can i do with this value?

File: cpp/Osmosis/ObjectStore/DirectoryValidator.h (hunk 4 above)
Comment (eliran-stratoscale) [inline cpp/Osmosis/ObjectStore/DirectoryValidator.h]: Done

**Comment by shlomi-stratoscale** [comment]
pushed after tiny fix.

**Pull request closed without merging**